notebooks/06a Modelling - classification.ipynb

```diff
@@ -0,0 +1,835 @@
+    "\n",
+    "e.g.\n",
+    "\n",
+    "class|count|weight\n",
+    "---|---|---\n",
+    "1|1000|0.1\n",
+    "2|100|1\n",
+    "3|500|0.2"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "weights = X_train.risk.value_counts().min() / X_train.risk.value_counts().sort_index()\n",
+    "\n",
+    "## Drop risk from training/test set\n",
+    "X_train.drop(columns=\"risk\", inplace=True)\n",
+    "X_test.drop(columns=\"risk\", inplace=True)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Modeling\n",
+    "\n",
+    "Strategy is to try a number of classification models with:\n",
+    "\n",
+    "* GridsearchCV for hyperparameter tuning with cross validation, refitting full training set for best model\n",
+    "* Test all final models against the held-out test set.\n",
+    "* Explore feature importance of best performing model\n",
+    "* Explore fairness (next notebook) of best performing model\n",
+    "\n",
+    "Logistic regression models are excluded due to significant multi-collinearity (see analysis in Regression notebook)."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "gather": {
+     "logged": 1652705481582
+    }
+   },
+   "outputs": [],
+   "source": [
+    "# Initiate empty models dictionary\n",
+    "models = {}"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Prior model\n",
+    "\n",
+    "The simplest baseline model takes the most frequent class label as its prediction"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "gather": {
+     "logged": 1652705486989
+    }
+   },
+   "outputs": [],
+   "source": [
+    "model_name = \"prior\"\n",
+    "\n",
+    "# define gridsearch parameters\n",
+    "gsc = GridSearchCV(\n",
+    "    estimator=DummyClassifier(strategy=\"prior\"),\n",
+    "    param_grid={},\n",
+    "    cv=5,\n",
+    "    scoring=\"balanced_accuracy\",\n",
+    "    verbose=1,\n",
+    "    n_jobs=-1,\n",
+    "    refit=True,\n",
+    ")\n",
+    "\n",
+    "# takes ~1 second to run on a STANDARD_DS3_V2\n",
+    "models[model_name] = train_model(gsc, X_train, y_train)\n",
+    "models[model_name]"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Decision tree classifier (weighted)\n",
+    "\n",
+    "Simplest tree classifier using one tree"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "model_name = \"decisiontree\"\n",
+    "\n",
+    "# define gridsearch parameters\n",
+    "gsc = GridSearchCV(\n",
+    "    estimator=DecisionTreeClassifier(class_weight=weights.to_dict()),\n",
+    "    param_grid={\"max_depth\": [5, 10, None]},\n",
+    "    cv=5,\n",
+    "    scoring=\"accuracy\",\n",
```

Comment:
Should this be accuracy? You've used class weights but the dataset is still imbalanced so I'd expect this metric to be deceptively high.

Reply:
Excellent catch. Will update